impeller/geometry/path.cc

```diff
@@ -13,12 +13,46 @@
 
 namespace impeller {
 
-Path::Path() {
+Path::Path() : copy_on_shared_write_data_(new Data()) {
```

Comment:
The way that Paths are structured today, the mutating methods are only visible to PathBuilder, and the rest of Impeller treats them as immutable. This could be implemented differently: we could have path builder hold the data internally and then construct path with all data in the ctor without hidden mutation.

Reply:
What I mean is: I think this code will be much simpler if you treat the shared data as read-only and instead adjust how PathBuilder constructs paths.

Reply:
Ah, PathBuilder could hold an instance of the Data struct instead of an instance of Path? Then only it would need a "shared-lock" variable, the Path would be blissfully ignorant of whether its data was being shared with a PathBuilder...?

Reply:
I think on construction the PathBuilder would move the data into the Path and then there would never be any need for locking

Reply:
I see. It could also trim the data as it moved it which would be a plus.

Right now we don't need another copy to call TakePath() unless the caller keeps using the PathBuilder.

Reply:
I guess this is no different than DisplayListBuilder and DisplayList where it trims the storage and hands it off. And, it's way cleaner in the long run as it isolates mutation code explicitly from an immutable representation.

Reply:
Done. Path is now truly immutable and cloned data is truncated (naturally by the default copy constructor).